packages/contracts/src/contracts/anchor.ts

```diff
@@ -110,26 +107,46 @@ export class AnchorContract {
       .sort((a, b) => a.args.leafIndex - b.args.leafIndex) // Sort events in chronological order
       .map((e) => e.args.commitment);
 
-    const commitments = [...storedContractInfo.leaves, ...newCommitments];
-
-    // extract the commitments from the events, and update the storage
-    await this.mixersInfo.setMixerStorage(this._contract.address, currentBlock, commitments);
-
-    return commitments;
+    return {
+      lastQueriedBlock: currentBlock,
+      newLeaves: newCommitments,
+    };
   }
 
   /*
    * Generate Merkle Proof
    *  This will
-   *  1- Fetch leaves
-   *  2- build the merkle tree & get the commitment leaf index
-   * 	3- Get the Merkle Path
+   *  1- Create the merkle tree with the leaves in local storage
+   *  2- Fetch the missing leaves
+   *  3- Insert the missing leaves
+   *  4- Compare against historical roots before adding to local storage
+   *  5- return the path to the leaf.
    * */
   async generateMerkleProof(deposit: Deposit) {
-    const leaves = await this.getDepositLeaves();
-    logger.trace(`Leaves ${leaves.length}`, leaves);
-    const tree = new MerkleTree('eth', 20, leaves);
-    let leafIndex = leaves.findIndex((commitment) => commitment == bufferToFixed(deposit.commitment));
+    const storedContractInfo = await this.mixersInfo.getMixerStorage(this._contract.address);
+    const tree = new MerkleTree('eth', 20, storedContractInfo.leaves);
```

Comment:
Can we query the contract for the merkle tree height as well? Currently 20 is hardcoded and should be swapped for a definitely foolproof value.

Reply:
Added this